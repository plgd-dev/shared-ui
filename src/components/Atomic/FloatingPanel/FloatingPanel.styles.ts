import { css } from '@emotion/react'

import { ThemeType, getThemeColor } from '../_theme'

export const referenceItem = (theme: ThemeType) => css`
    display: flex;
    align-items: center;
    color: ${getThemeColor(theme, `FloatingPanel.Icon.color`)};
    transition: all 0.3s;
    cursor: pointer;

    &:hover {
        color: ${getThemeColor(theme, `FloatingPanel.Icon.hover.color`)};
    }
`

export const activeReferenceItem = (theme: ThemeType) => css`
    color: ${getThemeColor(theme, `FloatingPanel.Icon.hover.color`)};
`

export const floatingPanel = (theme: ThemeType) => css`
    z-index: 10;
    background: ${getThemeColor(theme, `FloatingPanel.Content.background`)};
    border: 1px solid ${getThemeColor(theme, `FloatingPanel.Content.borderColor`)};
    box-shadow: 0 30px 40px rgba(28, 52, 99, 0.1);
    border-radius: 8px;
    min-width: 400px;
    max-width: 600px;
`

export const content = css`
    padding: 0;
`
