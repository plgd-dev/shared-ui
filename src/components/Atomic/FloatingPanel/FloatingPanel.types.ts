import { ReactNode } from 'react'

export type Props = {
    children: ReactNode
    reference: any
    title?: string
}
