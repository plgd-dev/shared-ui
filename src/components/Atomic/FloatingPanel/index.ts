export { default } from './FloatingPanel'
export * from './FloatingPanel'
