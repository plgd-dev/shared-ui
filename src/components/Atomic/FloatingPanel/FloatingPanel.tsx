import { cloneElement, FC, useEffect, useRef, useState } from 'react'
import { offset, shift, useFloating } from '@floating-ui/react'
import { motion, AnimatePresence } from 'framer-motion'
import { FloatingPortal } from '@floating-ui/react-dom-interactions'

import { Props } from './FloatingPanel.types'
import * as styles from './FloatingPanel.styles'
import { hasEventBlocker } from '../_utils/envets'

const FloatingPanel: FC<Props> = (props) => {
    const { children, reference, title } = props
    const [open, setOpen] = useState(false)
    const ref = useRef(null)
    const { x, y, refs, strategy } = useFloating({
        placement: 'bottom-end',
        strategy: 'fixed',
        middleware: [shift(), offset(4)],
    })

    useEffect(() => {
        function handleClickOutside(event: any) {
            // @ts-ignore
            if (ref?.current && !ref?.current?.contains(event.target) && !hasEventBlocker(event, 'data-block-outside-click')) {
                setOpen(false)
            }
        }
        if (open) {
            document.addEventListener('mousedown', handleClickOutside)
            return () => {
                document.removeEventListener('mousedown', handleClickOutside)
            }
        }
    }, [ref, open])

    return (
        <div>
            <span
                css={[styles.referenceItem, open && styles.activeReferenceItem]}
                data-block-outside-click={true}
                onClick={() => setOpen(!open)}
                ref={refs.setReference}
            >
                {cloneElement(reference)}
            </span>
            <FloatingPortal>
                <AnimatePresence>
                    {open && (
                        <motion.div
                            animate={{ opacity: 1, scale: 1, y: 0 }}
                            css={styles.floatingPanel}
                            exit={{ opacity: 0, y: 0 }}
                            initial={{ opacity: 0, scale: 0.85, y: -100 }}
                            ref={refs.setFloating}
                            style={{
                                position: strategy,
                                top: y ?? 0,
                                left: x ?? 0,
                                width: 'max-content',
                            }}
                            transition={
                                // When in "grouped phase", make the transition faster
                                // The open delay becomes 1ms during this phase.
                                { type: 'spring', damping: 20, stiffness: 300 }
                            }
                        >
                            <div ref={ref}>
                                {title ? (
                                    <div css={styles.header}>
                                        <div css={styles.headline}>{title}</div>
                                    </div>
                                ) : null}
                                <div css={styles.content}>{children}</div>
                            </div>
                        </motion.div>
                    )}
                </AnimatePresence>
            </FloatingPortal>
        </div>
    )
}

FloatingPanel.displayName = 'FloatingPanel'

export default FloatingPanel
