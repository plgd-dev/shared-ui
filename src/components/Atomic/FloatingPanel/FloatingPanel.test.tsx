import { render } from '../../../../test/jest-wrapper'
import FloatingPanel from './FloatingPanel'

describe('<FloatingPanel>', () => {
    it('renders without crashing', () => {
        const { container, asFragment } = render(<FloatingPanel reference={<div>Reference</div>}>Test</FloatingPanel>)
        expect(container).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
    })
})
