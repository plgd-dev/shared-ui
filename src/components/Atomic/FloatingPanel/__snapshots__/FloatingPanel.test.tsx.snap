// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FloatingPanel> renders without crashing 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #81868C;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  cursor: pointer;
}

.emotion-0:hover {
  color: #2261AE;
}

<div>
    <span
      class="emotion-0"
      data-block-outside-click="true"
    >
      <div>
        Reference
      </div>
    </span>
  </div>
</DocumentFragment>
`;
